deckhand/engine/secret_substitution.py

```diff
@@ -0,0 +1,157 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import yaml
+
+import jsonschema
+
+from deckhand.engine.schema.v1_0 import default_schema
+from deckhand import errors
+
+
+class SecretSubstitution(object):
+    """Class for secret substitution logic for YAML files.
+
+    This class is responsible for parsing, validating and retrieving secret
+    values for values stored in the YAML file. Afterward, secret values will be
+    substituted or "forward-repalced" into the YAML file. The end result is a
+    YAML file containing all necessary secrets to be handed off to other
+    services.
+
+    :param data: YAML data that requires secrets to be validated, merged and
+        consolidated.
+    """
+
+    def __init__(self, data):
+        try:
+            self.data = yaml.safe_load(data)
+        except yaml.YAMLError:
+            raise errors.InvalidFormat(
+                'The provided YAML file cannot be parsed.')
+
+        self.validate_data()
+
+    class SchemaVersion(object):
+        """Class for retrieving correct schema for pre-validation on YAML.
+
+        Retrieves the schema that corresponds to "apiVersion" in the YAML
+        data. This schema is responsible for performing pre-validation on
+        YAML data.
+        """
+
+        schema_versions_info = [{'version': 'v1', 'schema': default_schema}]
+
+        def __init__(self, schema_version):
+            self.schema_version = schema_version
+
+        @property
+        def schema(self):
+            return [v['schema'] for v in self.schema_versions_info
+                    if v['version'] == self.schema_version][0].schema
+
+    def validate_data(self):
+        """Pre-validate that the YAML file is correctly formatted.
+
+        The YAML file must adhere to the following bare minimum format:
+
+        .. code-block:: yaml
+
+            ---
+            apiVersion: service/v1
+            kind: ConsumerOfCertificateData
+            metadata:
+              substitutions:
+                - dest: .tls_endpoint.certificate
+                  src:
+                    apiVersion: deckhand/v1
+                    kind: Certificate
+                    name: some-certificate-asdf-1234
+                # Forward-reference to specific section under "data" below.
+                - dest: .tls_endpoint.certificateKey
+                  src:
+                    apiVersion: deckhand/v1
+                    kind: CertificateKey
+                    name: some-certificate-key-asdf-1234
+            data:
+              tls_endpoint:
+                  certificate: null  # Data to be substituted.
+                  certificateKey: null  # Data to be substituted.
+        """
+        self._validate_with_schema()
+
+        # Validate that each "dest" field exists in the YAML data.
+        substitutions = self.data['metadata']['substitutions']
```

Comment:
We should just insert data into the document based on the dest field (even if that is means inserting a narrow, deep object) and then allow the final schema validation determine if that dest field is valid.